letsencrypt-nginx/letsencrypt_nginx/nginxparser.py

```diff
@@ -1,56 +1,241 @@
+                result = result + self._read()
+
+        def _parse_string(self):
+            startVal = self._read()
+            result = startVal
+            while True:
+                val = self._read()
+                result = result + val
+                if val == None:
+                    raise ParseException("Invalid configuration file, unfinished string literal")
+                if val == startVal:
+                    return result
+
+        def _parse_value(self):
+            self._pass_whitespace(True)
+            result = ''
+            while True:
+                val = self._peek()
+                if val is None or val in ['{', ';']:
+                    result = result.rstrip()
+                    if result == '':
+                        return None
+                    return result
+                if val == '"' or val == "'":
+                    result = result + self._parse_string()
+                else:
+                    result = result + self._read()
+
+        def _parse_modifier(self):
+            """Parses a modifier - this is slightly looser than the original parser"""
+            self._pass_whitespace(True)
+            val = self._peek()
+            if val in ['=', '~', '^']:
+                first = self._read()
+                second = self._peek()
+                if not self._is_whitespace(second, True):
+                    self._read()
+                    return first + second
+                else:
+                    return first
+            return None
+
+        def _expect(self, val):
+            """Reads a character, and raises an exception if it's not the expected value"""
+            r = self._read()
+            if r != val:
+                raise ParseException("Expected %s but got %s" % (val, r))
+
+        def _finish_assignment(self, directive, val):
+            self._expect(';')
+            directive.append(val)
+            self._pass_whitespace(False)
+
+            comment = self._parse_comment()
+            if comment:
+                directive.append(''.join(comment))
+            else:
+                self._parse_newline()
+            return directive
+
+        def _finish_block(self, directive, val):
+            self._expect('{')
+            self._parse_newline()
+            block = self._parse_block()
+            self._expect('}')
+            self._parse_newline()
+            if val:
+                directive.append(val)
+            return [directive, block]
+
+        def _parse_assignment_or_block(self):
+            key = self._parse_key()
+            if key is None:
+                return None
+            result = [key]
+            if key == 'location':
+                mod = self._parse_modifier()
+                if mod is not None:
+                    result.append(mod)
+            val = self._parse_value()
+            nx = self._peek()
+
+            if nx == ';':
+                return self._finish_assignment(result, val)
```

Comment:
Yes, I noted this, but I didn't want to change it in this PR since some of that awkwardness was there from before (several type-checks that actually differentiate different directive types). I would be fine with fixing it, but it probably fits better as a self contained refactoring.